### PR TITLE
Add RelBench integration for PyTorch Geometric GNN+LLM applications

### LINKED ISSUE #9839: Integrating GNNs and LLMs for Enhanced Data Warehouse Understanding and Lineage Analysis
### 🚀 The feature, motivation and pitch

@puririshi98
**The Feature, Motivation, and Pitch**:  
Modern data warehouses face several critical challenges, including understanding data lineage, identifying data silos, and interpreting complex transformations in ETL processes. Existing systems, including those leveraging LLMs, fall short of addressing these challenges due to a lack of grounding in the structured relationships inherent in data warehouses.

I propose a feature to enable seamless integration of **Graph Neural Networks (GNNs)** and **Large Language Models (LLMs)** to model data warehouses as graphs, allowing for improved reasoning and understanding of:  
- **Data Lineage**: By representing transformations, dependencies, and data flow as graph structures, a hybrid GNN+LLM system can analyze and explain lineage paths.  
- **Data Silos**: Detecting disconnected components in the data warehouse graph to suggest potential integrations.  
- **ETL Transformations**: Providing insights into how raw data evolves through complex transformations into actionable insights.  
- **Schema and Query Understanding**: Modeling schemas as graphs can improve LLM capabilities to generate and interpret queries based on relationships between tables.  

The integration would involve using PyG (PyTorch Geometric) for GNN modeling and extending existing libraries for training hybrid GNN+LLM architectures. This will allow data warehouse systems to gain both structural awareness (from GNNs) and semantic reasoning (from LLMs), reducing hallucinations and improving the interpretability of predictions and queries.

### Alternatives

- Existing LLM solutions provide semantic reasoning but often hallucinate without structured context.
- Pure GNN solutions focus on structural reasoning but lack the language capabilities needed for intuitive query interaction.
- Ensemble systems attempt to combine these capabilities but lack a unified framework for data warehouse tasks.

### Additional context

- Data lineage tools (e.g., Neo4j integrations or metadata graphing systems) could serve as a starting point for the graph representation of data warehouses.
- Recent work on combining GNNs with LLMs for question answering and recommendation systems could provide foundational knowledge for this hybrid architecture.

This feature would enable the PyTorch Geometric community to explore real-world applications in data science, bridging the gap between NLP and data engineering.

### PR DESCRIPTION
**Pull request by AJamal27891**
# Add warehouse intelligence system with RelBench integration

**Closes #9839**

This PR implements a warehouse intelligence system for PyTorch Geometric, providing RelBench dataset integration and graph-based warehouse analysis capabilities using G-Retriever architecture for multi-task learning on data lineage, silo detection, and quality assessment tasks.

## Key Changes

**New Files:**
- `torch_geometric/datasets/relbench.py` - RelBench to HeteroData conversion utilities
- `torch_geometric/utils/data_warehouse.py` - Warehouse intelligence with G-Retriever integration
- `examples/llm/whg_demo.py` - Warehouse intelligence demonstration
- `test/datasets/test_relbench.py` - Comprehensive RelBench functionality tests (9 tests)
- `test/utils/test_data_warehouse.py` - Warehouse intelligence system tests (13 tests)

**Updated Files:**
- `torch_geometric/datasets/__init__.py` - Export RelBench utilities
- `pyproject.toml` - Optional dependency groups for `relbench` and `whg`

### RelBench Integration
- `create_relbench_hetero_data()` - Convert RelBench datasets to PyG HeteroData
- `RelBenchDataset` - PyG dataset wrapper for RelBench data
- `HeuristicLabeler` - Generate warehouse task labels from graph structure
- `RelBenchProcessor` - Process RelBench data with semantic embeddings

### Warehouse Intelligence System
- `WarehouseGRetriever` - G-Retriever architecture for warehouse analysis
- `WarehouseTaskHead` - Multi-task prediction (lineage, silo, quality)
- `WarehouseConversationSystem` - Natural language interface for warehouse queries
- `SimpleWarehouseModel` - Lightweight model for basic warehouse operations

### Multi-task Learning
- **Lineage prediction** - Trace data flow and dependencies
- **Silo detection** - Identify isolated data components
- **Quality assessment** - Detect anomalies and data quality issues

### Basic RelBench Integration
```python
from torch_geometric.datasets.relbench import create_relbench_hetero_data

# Convert RelBench dataset to PyG format
hetero_data = create_relbench_hetero_data(
    dataset_name='rel-f1',
    sample_size=100,
    create_lineage_labels=True,
    create_silo_labels=True,
    create_anomaly_labels=True
)
```

### Warehouse Intelligence System
```python
from torch_geometric.utils.data_warehouse import create_warehouse_demo

# Create warehouse conversation system
warehouse_system = create_warehouse_demo()

# Query warehouse intelligence
result = warehouse_system.process_query(
    "What is the data lineage in this warehouse?", 
    graph_data
)
print(result['answer'])
```

## Installation

```bash
# RelBench integration
pip install torch-geometric[relbench]

# Warehouse intelligence system
pip install torch-geometric[whg]

# Both features
pip install torch-geometric[relbench,whg]
```

## Testing

**Test Coverage:** 22 tests across 2 files, all passing ✅
- `test/datasets/test_relbench.py` - RelBench integration tests (9 tests)
- `test/utils/test_data_warehouse.py` - Warehouse intelligence tests (13 tests)

Includes comprehensive coverage of core functionality, edge cases, and error handling.

## Technical Implementation

The system integrates G-Retriever architecture with RelBench datasets to provide warehouse intelligence capabilities. Key technical features:

- **Semantic embeddings** using sentence transformers for text-based node features
- **Multi-task learning** with shared GNN backbone and task-specific heads
- **Heuristic labeling** for automatic warehouse task label generation
- **LLM integration** with fallback to traditional GNN approaches
- **Modular design** supporting both standalone and integrated usage

## Changelog Entry

Added to `CHANGELOG.md` under version 2.7.0:
```
- Added RelBench integration with data warehouse lineage tasks (#10353)
```

This entry covers the complete warehouse intelligence system implementation including RelBench integration, G-Retriever architecture, multi-task learning capabilities, and comprehensive testing.

### CODE REVIEW

**Comment by puririshi98** [comment]
please add a changelog entry

**Comment by puririshi98** [comment]
please also address linting, will do a deep review this week

**Comment by puririshi98** [comment]
there is alot of overlap with this existing example: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/rdl.py
I was under the impression that you were going to integrate this into a pipeline similar to G-retriever (see https://github.com/pyg-team/pytorch_geometric/tree/master/examples/llm) where you could "talk to your data warehouse" (since G-retriever style GNN+LLM enables "talk to your graph"). please align your API's with the existing examples/rdl from the core contributors and let me know when you have a working "talk to your data warehouse"  example. 
See my talk here if you'd like more details about "talk to your graph"
https://www.devreal.ai/graph-exchange-may-2025/

**Review by puririshi98 — CHANGES_REQUESTED** [review]
see above comment

**Comment by AJamal27891** [comment]
Hi @puririshi98,

The PR has been updated to address your feedback:

**Overlap resolved** - Removed duplicate examples and aligned with existing examples/rdl.py patterns rather than conflicting with them. The new WHG-Retriever provides complementary warehouse-specific functionality.

**Conversational interface implemented** - "Talk to your data warehouse" working as requested, following the G-Retriever style you demonstrated in your Graph Exchange talk.

**API alignment complete** - Now follows existing examples/llm patterns and integrates with PyG infrastructure.

**Technical fixes addressed** - Changelog entry added, linting resolved, proper code organization.

The system provides GAT-based warehouse analysis with multi-task learning and natural language interface, building on PyG's existing capabilities rather than duplicating them.

Ready for your review - thank you for the guiding comments.

File: test/utils/test_relbench.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+"""Tests for RelBench Integration Utilities.
+
+This test suite uses local fixtures to avoid network dependencies in CI.
+Tests basic functionality of RelBench integration utilities.
+"""
+
+# Import the module under test
+try:
+    from torch_geometric.datasets.relbench import get_warehouse_task_info
```
Comment (puririshi98) [inline test/utils/test_relbench.py]: if we are testing torch_geometric.datasets.relbench then this test should fall under `test/datasets/test_relbench.py`. please move this file accordingly

File: CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added the `Teeth3DS` dataset, an extended benchmark for intraoral 3D scan analysis ([#9833](https://github.com/pyg-team/pytorch_geometric/pull/9833))
 - Added `torch.device` to `PatchTransformerAggregation` [#10342](https://github.com/pyg-team/pytorch_geometric/pull/10342)
 - Added `torch.device` to normalization layers [#10341](https://github.com/pyg-team/pytorch_geometric/pull/10341)
+- Added RelBench integration with data warehouse lineage tasks ([#10353](https://github.com/pyg-team/pytorch_geometric/pull/10353))
```
Comment (puririshi98) [inline CHANGELOG.md]: move this to the top of changelog since everything above it has already been added

File: examples/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+# Examples package
```
Comment (puririshi98) [inline examples/__init__.py]: you should not be makign examples into a package, please remove this

File: examples/llm/__init__.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+"""PyTorch Geometric LLM examples.
+
+Graph-based retrieval and conversation utilities for PyG.
+"""
+
+# WHG-Retriever components
+try:
+    from .whg_model import WarehouseGRetriever, WarehouseTaskHead  # noqa: F401
```
Comment (puririshi98) [inline examples/llm/__init__.py]: again, examples should be standalone scripts, not modules. please move this stuff under the torch_geometric folder. Maybe these can go under torch_geometric/utils/data_warehouse.py.

(also open to suggestions on the naming of the utils file)

File: examples/llm/relbench_warehouse_demo.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+and G-Retriever preparation for future LLM integration.
+
+Requirements:
+`pip install relbench[full] sentence-transformers`
+
+Paper references:
+- RelBench: https://arxiv.org/abs/2407.20060
+- G-Retriever: https://arxiv.org/abs/2402.07630
+"""
+import argparse
```
Comment (puririshi98) [inline examples/llm/relbench_warehouse_demo.py]: what is the different between this file and whg_demo?

**Review by puririshi98 — CHANGES_REQUESTED** [review]
this looks REALLY cool! Please address my surface level concerns and I will try to do another review tomorrow

File: test/utils/test_relbench.py (hunk 1 above)
Comment (AJamal27891) [reply test/utils/test_relbench.py]: I have created 2 test files
1. `test/datasets/test_relbench.py `- Tests for RelBench dataset integration
2. `test/utils/test_data_warehouse.py `- Tests for warehouse utilities

File: examples/llm/relbench_warehouse_demo.py (hunk 5 above)
Comment (AJamal27891) [reply examples/llm/relbench_warehouse_demo.py]: Removed ! since it was only demonstrating the relbench integration

File: examples/llm/whg_demo.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
+"""Warehouse Intelligence Demo - Out of the Box Usage.
+
+Demo of PyTorch Geometric warehouse intelligence with RelBench data.
+Includes lineage detection, silo analysis, and quality assessment.
+
```
Comment (puririshi98) [inline examples/llm/whg_demo.py:5]: looks good so far but please update the examples/llm/readmME.md and ill take a deeper look later today

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/llm/whg_demo.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+Usage:
+    python examples/llm/whg_demo.py
```
Comment (puririshi98) [inline examples/llm/whg_demo.py]: can you run this with https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pyg/ 25.05 and make sure everything runs. Please attach a log of that run so that I can review

**Review by puririshi98 — CHANGES_REQUESTED** [review]

File: examples/llm/whg_demo.py (hunk 7 above)
Comment (AJamal27891) [reply examples/llm/whg_demo.py]: I spun everything up in nvcr.io/nvidia/pyg:25.05-py3 (inside the container, no local mods) and the demo runs end-to-end:

```bash
"docker exec -w /opt/pyg/pytorch_geometric 7ddc1ab2fd5f python examples/llm/whg_demo.py"
PyG SentenceTransformer requires tokenized input
PyG SentenceTransformer requires tokenized input
PyG SentenceTransformer requires tokenized input
PyG SentenceTransformer requires tokenized input
PyG SentenceTransformer requires tokenized input
PyG SentenceTransformer requires tokenized input
PyG SentenceTransformer requires tokenized input
PyG SentenceTransformer requires tokenized input
PyG SentenceTransformer requires tokenized input
/opt/pyg/pytorch_geometric/examples/llm/../../torch_geometric/nn/nlp/llm.py:97: UserWarning: LLM is being used on CPU, which may be slow
  warnings.warn("LLM is being used on CPU, which may be slow")
/opt/pyg/pytorch_geometric/examples/llm/../../torch_geometric/utils/data_warehouse.py:655: UserWarning: Converting a tensor with requires_grad=True to a scalar may lead to unexpected behavior.
Consider using tensor.detach() first. (Triggered internally at /opt/pytorch/pytorch/aten/src/ATen/native/Scalar.cpp:22.)
  confidence = probs.max(dim=-1).values.mean().item()
Warehouse Intelligence Demo
==============================

Step 1: Loading RelBench data
Loading Database object from /root/.cache/relbench/rel-f1/db...
Done in 0.06 seconds.
Loaded graph with 9 node types
Node types: ['standings', 'constructor_standings', 'circuits', 'constructor_results', 'qualifying', 'drivers', 'races', 'results', 'constructors']
Converted to homogeneous: 450 nodes, 236 edges

Step 2: Creating warehouse conversation system
Setting up 'TinyLlama/TinyLlama-1.1B-Chat-v0.1' with configuration: {'revision': 'main'}
Warehouse conversation system ready

Step 3: Running warehouse intelligence queries

--- Query 1: What is the data lineage in this warehouse? ---
Answer:  The data lineage in this warehouse can be traced as follows:

Direct connections: The raw data sources for this warehouse are spread across multiple data centers and data sources. The data lineage traces back to these sources.

Staged data: The intermediate processing steps involved in data transfo...

Quantitative Analysis: Unknown lineage detectedacross 450 entities (confidence: 0.206)
Query type: lineage
Confidence: N/A

--- Query 2: Are there any data silos? ---
Answer:  The data warehouse is composed of 450 entities, 236 relationships, and sparsely connected connections. The graph structure suggests a hierarchical structure with 450 nodes and 236 relationships. The data sources are isolated or poorly connected, with no bridges connecting different data domains. Cl...

Quantitative Analysis: Analytics: 0 isolated silosout of 450 entities (0.2% isolation rate)
Query type: silo
Confidence: N/A

--- Query 3: What is the data quality status? ---
Answer:  The data in this warehouse is in a highly inconsistent and unstructured format. The data is not cleaned, validated, or formatted in a consistent manner. The data is also not structured in a way that allows for easy analysis and querying.

The data is also not regularly updated or maintained. This m...

Quantitative Analysis: Analytics: Quality score 0.623(GOOD overall status)
Query type: quality
Confidence: N/A

--- Query 4: Analyze the impact of changes in this warehouse ---
Answer:  The warehouse is a large-scale relational database with multiple entity types. The graph structure is comprised of 450 entities, 236 relationships. The connectivity is sparse, with an average degree of 1.0. The impact analysis will focus on the downstream effects of data changes, considering high-i...

Quantitative Analysis: Analytics: 0 high-impact entitiesdetected (LOW risk)
Query type: impact
Confidence: N/A

Step 4: Conversation History
------------------------------
1. Q: Are there any data silos?...
   A:  The data warehouse is composed of 450 entities, 236 relationships, and sparsely...
2. Q: What is the data quality status?...
   A:  The data in this warehouse is in a highly inconsistent and unstructured format....
3. Q: Analyze the impact of changes in this warehouse...
   A:  The warehouse is a large-scale relational database with multiple entity types. ...

Demo completed. Processed 4 queries total.

Features demonstrated:
- RelBench data integration
- Multi-task warehouse intelligence
- Natural language query processing
- Lineage, silo, and quality analysis
```

The demo runs out of the box no modifications just the command `python examples/llm/whg_demo.py`  
the output shows the integration between `rel-bench` and the `data warehouse`

Comment (puririshi98) [reply examples/llm/whg_demo.py]: curious why does this cut off:
--- Query 1: What is the data lineage in this warehouse? ---
```
Answer:  The data lineage in this warehouse can be traced as follows:

Direct connections: The raw data sources for this warehouse are spread across multiple data centers and data sources. The data lineage traces back to these sources.

Staged data: The intermediate processing steps involved in data transfo...

Quantitative Analysis: Unknown lineage detectedacross 450 entities (confidence: 0.206)
```

i notice alot of sentences are truncated. please help me understand why or share the full logs. otherwise this looks really cool

**Comment by puririshi98** [comment]
can you please make CI green as well?

File: examples/llm/whg_demo.py (hunk 7 above)
Comment (AJamal27891) [reply examples/llm/whg_demo.py:7]: Is just me keeping the demo log short.
I cap the LLM at a small token budget for fast inference:
<br>
```python
llm_response = model.inference(
    question=[contextual_prompt], 
    x=x, edge_index=edge_index,
    batch=batch, 
    max_out_tokens=150  # ← limiting the llm output.
)
````

**Comment by puririshi98** [comment]
please make sure CI green and you add fineutning to the example

**Comment by AJamal27891** [comment]
> please make sure CI green and you add fineutning to the example

[whg_phi3_train_log_concise_clean.txt](https://github.com/user-attachments/files/21770594/whg_phi3_train_log_concise_clean.txt)


this file contains the full log with 1 epoch training example using phi3

File: test/datasets/test_relbench.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+            RelBenchProcessor,
+            create_relbench_hetero_data,
+        )
+
+        # Test that classes can be imported
+        assert RelBenchDataset is not None
+        assert RelBenchProcessor is not None
+        assert create_relbench_hetero_data is not None
+    except ImportError:
+        pytest.skip("RelBench not available")
```
Comment (wsad1) [inline test/datasets/test_relbench.py]:
```suggestion
from torch_geometric.datasets.relbench import (
    RelBenchDataset,
    RelBenchProcessor,
    create_relbench_hetero_data,
)
```
No need to check this, the test will only be run on a version that has the `relbench.py` file.

File: test/datasets/test_relbench.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            pytest.skip("Sentence transformers not available in CI")
+        else:
+            raise
+
+
+def test_create_relbench_hetero_data() -> None:
+    """Test create_relbench_hetero_data function."""
+    try:
+        from torch_geometric.datasets.relbench import \
+            create_relbench_hetero_data  # noqa: E501
```
Comment (wsad1) [inline test/datasets/test_relbench.py]: Same as above, no need to put these imports in a Try catch block.

File: torch_geometric/utils/data_warehouse.py [hunk 10]
```diff
@@ -0,0 +1,1886 @@
...
+except ImportError:
+    HAS_GRETRIEVER = False
+
+    class GRetriever:  # type: ignore
+        def __init__(self, *args: Any, **kwargs: Any) -> None:
+            raise ImportError("GRetriever requires PyG with LLM support")
+
+    class LLM:  # type: ignore
+        def __init__(self, *args: Any, **kwargs: Any) -> None:
+            raise ImportError("LLM requires PyG with LLM support")
```
Comment (wsad1) [inline torch_geometric/utils/data_warehouse.py:105]:
```suggestion
from torch_geometric.nn.models import GRetriever
from torch_geometric.nn.nlp import LLM
HAS_GRETRIEVER = True
```
Why do we need this try catch block. GRetriever, is part of `torch_geometric` and it was added before `data_warehouse.py`, so no need to check for this. 
We only need to have such try catch blocks for optional external dependencies.

File: torch_geometric/datasets/relbench.py [hunk 11]
```diff
@@ -0,0 +1,934 @@
...
+    """Create HeteroData from RelBench dataset with warehouse enhancements.
+
+    Args:
+        dataset_name: Name of the RelBench dataset
+        sbert_model: Sentence transformer model for embeddings
+        sample_size: Optional sample size for large datasets
+        create_lineage_labels: Whether to create lineage labels
+        create_silo_labels: Whether to create silo labels
+        create_anomaly_labels: Whether to create anomaly labels
+        external_labels: Pre-computed labels to use instead of heuristics
```
Comment (wsad1) [inline torch_geometric/datasets/relbench.py:614]: Could you add more details about these labels. 
Is there a paper one could look at to understand how to use these labels. What is the shape of `external_labels`.

**Review by wsad1 — CHANGES_REQUESTED** [review]
Thanks a lot for adding this. 
Left some initial comments. Will continue to review since this is a large PR.

**Review by puririshi98 — CHANGES_REQUESTED** [review]
one quick change, move the utils changes into torch_Geometric.llm instead, ping me on slack to quick review once you do

File: torch_geometric/datasets/relbench.py [hunk 12]
```diff
@@ -0,0 +1,978 @@
...
+            create_silo_labels=create_silo_labels,
+            create_anomaly_labels=create_anomaly_labels)
+
+    hetero_data.dataset_name = 'dummy_warehouse_data'
+    hetero_data.embedding_dim = EMBEDDING_DIM
+
+    return hetero_data
+
+
+class RelBenchDataset(InMemoryDataset):
```
Comment (wsad1) [inline torch_geometric/datasets/relbench.py:717]: Lets reuse functionality from relbench for creating/downloading this dataset
See here https://github.com/pyg-team/pytorch_geometric/blob/6208605b4d9711b86c7f06031b17efc5ea445afb/examples/rdl.py#L625 for more details.

File: torch_geometric/datasets/relbench.py (hunk 11 above)
Comment (wsad1) [reply torch_geometric/datasets/relbench.py:614]: move this to LLM folder under utility section. 
Also is there some paper or blog that explain this so that our community understand this better?

File: torch_geometric/datasets/relbench.py [hunk 13]
```diff
@@ -0,0 +1,978 @@
...
+
+
+class RelBenchError(Exception):
+    """Custom exception for RelBench-related errors."""
+
+
+class HeuristicLabeler:
+    """Generate warehouse task labels from foreign keys and graph structure."""
+    def __init__(self) -> None:
+        self.use_networkx = NETWORKX_AVAILABLE
```
Comment (wsad1) [inline torch_geometric/datasets/relbench.py]: This isn't used anywhere.

File: torch_geometric/datasets/relbench.py [hunk 14]
```diff
@@ -0,0 +1,978 @@
...
+except ImportError:
+    logger.warning("PyG graph analysis not available, using basic heuristics")
+    GRAPH_ANALYSIS_AVAILABLE = False
+
+try:
+    import networkx as nx
+    NETWORKX_AVAILABLE = True
+except ImportError:
+    logger.warning("NetworkX not available, using simplified graph analysis")
+    NETWORKX_AVAILABLE = False
```
Comment (wsad1) [inline torch_geometric/datasets/relbench.py]: Drop the `try except`, its now doing anything useful. 
The user will just install these if they are not preset. Do this in other places too where possible. 
The `try except` should only be there is there is a useful fallback.

**Review by wsad1 — COMMENTED** [review]
Lets move the utils also to `torch_geometric/llm`, Also drop most of the `relbench.py` code and move it to utils, see comments for more details.

File: torch_geometric/datasets/relbench.py (hunk 12 above)
Comment (wsad1) [reply torch_geometric/datasets/relbench.py:717]: has this comment been addressed. why do we need a RelBenchDataset?

File: torch_geometric/llm/data_warehouse.py [hunk 15]
```diff
@@ -0,0 +1,1885 @@
...
+                      "Medium({medium}), High({high})\n"
+                      "• Risk assessment: {risk_level}")
+    }
+}
+
+# Set up logging
+logger = logging.getLogger(__name__)
+
+
+class GATWrapper(nn.Module):
```
Comment (wsad1) [inline torch_geometric/llm/data_warehouse.py:75]: why do we need a gat wrapper?

**Review by wsad1 — COMMENTED** [review]
@AJamal27891  Thanks for all the great work till now. Really appreciate your effort.
But any chance this PR can be broken down into smaller PRs. There are just too many things to review, and its taking too long which I know can be frustrating for you too. I think the best way to move faster is to break things down.

**Comment by puririshi98** [comment]
closing in lieu of new PRs https://github.com/pyg-team/pytorch_geometric/pull/10628
